README.md
# EventSource

This library is a pure JavaScript implementation of the [EventSource](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events) client. The API aims to be W3C compatible.

You can use it with Node.js.  Use the [original package](https://github.com/EventSource/eventsource) for the browser polyfill.

## Install

    npm install eventsource

## Extensions to the W3C API

### Adjusting reconnection interval

You can customize the interval between reconnections via the `reconnectionInterval` argument. It can also be disabled by setting it to `0`.

```javascript
var eventSourceInitDict = {reconnectionInterval: 5000};
var es = new EventSource(url, eventSourceInitDict);
```

### Setting HTTP request headers

You can define custom HTTP headers for the initial HTTP request. This can be useful for e.g. sending cookies
or to specify an initial `Last-Event-ID` value.

HTTP headers are defined by assigning a `headers` attribute to the optional `eventSourceInitDict` argument:

```javascript
var eventSourceInitDict = {headers: {'Cookie': 'test=test'}};
var es = new EventSource(url, eventSourceInitDict);
```

### Allow unauthorized HTTPS requests

By default, https requests that cannot be authorized will cause the connection to fail and an exception
to be emitted. You can override this behaviour, along with other https options:

```javascript
var eventSourceInitDict = {https: {rejectUnauthorized: false}};
var es = new EventSource(url, eventSourceInitDict);
```

Note that for Node.js < v0.10.x this option has no effect - unauthorized HTTPS requests are *always* allowed.

### HTTP status code on error events

Unauthorized and redirect error status codes (for example 401, 403, 301, 307) are available in the `status` property in the error event.

```javascript
es.onerror = function (err) {
  if (err) {
    if (err.status === 401 || err.status === 403) {
      console.log('not authorized');
    }
  }
};
```

### HTTP/HTTPS proxy

You can define a `proxy` option for the HTTP request to be used. This is typically useful if you are behind a corporate firewall.

```javascript
var es = new EventSource(url, {proxy: 'http://your.proxy.com'});
```


## License

MIT-licensed. See LICENSE


HISTORY.md
# 1.0.0

* First release, similar to v1.0.7 of original EventSource


src/eventsource.ts
import original from 'original';
import { parse } from 'url';
import { EventEmitter } from 'events';
import https from 'https';
import http from 'http';
import assert from 'assert';

export enum EventSourceReadyState {
  /** The connection has not yet been established, or it was closed and the user agent is reconnecting. */
  CONNECTING,
  /** The user agent has an open connection and is dispatching events as it receives them. */
  OPEN,
  /** The connection is not open, and the user agent is not trying to reconnect. Either there was a fatal error or the close() method was invoked. */
  CLOSED
}

export interface EventSourceInit {
  /**
   * Indicates whether the EventSource object was instantiated with CORS credentials set (true), or not (false, the default).
   *
   * @default false
   */
  withCredentials?: boolean;
  /**
   * The interval in milliseconds between reconnection attempts. Set to 0 to disable.
   *
   * @default 1000
   */
  reconnectionInterval?: number;
  /**
   * The headers to define in the HTTP request.
   */
  headers?: { [index: string]: string };
  /**
   * If HTTPS requests with invalid certificates should be rejected.
   *
   * @default true
   */
  rejectUnauthorized?: boolean;
  /**
   * A HTTP proxy to use in the request
   */
  proxy?: string;

  /** HTTPS options */
  https?: https.RequestOptions;
}

export default class EventSource extends EventEmitter {
  // Enums
  readonly CONNECTING = EventSourceReadyState.CONNECTING;
  readonly OPEN = EventSourceReadyState.OPEN;
  readonly CLOSED = EventSourceReadyState.CLOSED;
  static readonly CONNECTING = EventSourceReadyState.CONNECTING;
  static readonly OPEN = EventSourceReadyState.OPEN;
  static readonly CLOSED = EventSourceReadyState.CLOSED;

  // ready state
  private _readyState: EventSourceReadyState = EventSourceReadyState.CONNECTING;

  // URL
  private originalUrl: string;
  private _url: string;

  // Other options
  private _withCredentials: boolean;
  private reconnectionInterval: number;
  private headers: { [index: string]: string };
  private rejectUnauthorized: boolean;
  private proxy?: string;
  private httpsOptions?: https.RequestOptions;

  // event source vars
  private request?: http.ClientRequest;
  private connectionInProgress = false;
  private lastEventId = '';

  // chunk vars
  private isFirstChunk = false;
  private chunkBuffer?: Buffer;
  private discardTrailingNewline = false;
  private startingPosition = 0;
  private startingFieldLength = -1;

  // event vars
  private eventName = '';
  private data = '';

  /**
  * Creates a new EventSource object.
  *
  * @param url A string giving the URL that will provide the event stream.
  * @param [eventSourceInitDict] The EventStream options. See README for details.
  * @api public
  * */
  constructor(url: string, eventSourceInitDict: EventSourceInit = {}) {
    super();

    this.originalUrl = url;
    this._url = url;
    this._withCredentials = eventSourceInitDict.withCredentials ?? false;
    this.reconnectionInterval = eventSourceInitDict.reconnectionInterval ?? 1000;
    this.headers = {
      'Cache-Control': 'no-cache',
      Accept: 'text/event-stream',
      ...eventSourceInitDict.headers,
    };
    this.rejectUnauthorized = eventSourceInitDict.rejectUnauthorized ?? true;
    this.proxy = eventSourceInitDict.proxy;
    this.httpsOptions = eventSourceInitDict.https;

    if (this.headers['Last-Event-ID']) {
      this.lastEventId = this.headers['Last-Event-ID'];
      delete this.headers['Last-Event-ID'];
    }

    process.nextTick(() => this.connect());
  }

  // eslint-disable-next-line class-methods-use-this
  private connect(): void {
    let reqOptions: https.RequestOptions = parse(this._url);
    let isSecure = reqOptions.protocol === 'https:';
    reqOptions.headers = this.headers;
    if (this.lastEventId) {
      reqOptions.headers['Last-Event-ID'] = this.lastEventId;
    }

    // Legacy: this should be specified as `eventSourceInitDict.https.rejectUnauthorized`,
    // but for now exists as a backwards-compatibility layer
    reqOptions.rejectUnauthorized = this.rejectUnauthorized;

    // If specify http proxy, make the request to sent to the proxy server,
    // and include the original url in path and Host headers
    if (this.proxy) {
      const proxy = parse(this.proxy);
      isSecure = proxy.protocol === 'https:';

      reqOptions.protocol = isSecure ? 'https:' : 'http:';
      reqOptions.path = this._url;
      reqOptions.headers.Host = reqOptions.host ?? undefined;
      reqOptions.hostname = proxy.hostname;
      reqOptions.host = proxy.host;
      reqOptions.port = proxy.port;
    }

    // If https options are specified, merge them into the request options
    if (this.httpsOptions) {
      reqOptions = {
        ...reqOptions,
        ...this.httpsOptions,
      };
    }

    if (this._withCredentials) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      reqOptions.withCredentials = this._withCredentials;
    }

    this.request = (isSecure ? https : http).request(reqOptions, (res) => {
      this.connectionInProgress = false;
      // Handle HTTP errors
      if (res.statusCode && res.statusCode >= 500) {
        this.emit('error', { type: 'error', status: res.statusCode, message: res.statusMessage });
        this.onConnectionClosed();
        return;
      }

      // Handle HTTP redirects
      if (res.statusCode === 301 || res.statusCode === 302 || res.statusCode === 307) {
        if (!res.headers.location) {
          // Server sent redirect response without Location header.
          this.emit('error', { type: 'error', status: res.statusCode, message: 'Server sent redirect response without Location header.' });
          return;
        }
        if (res.statusCode === 301) this.originalUrl = res.headers.location;
        this._url = res.headers.location;
        process.nextTick(() => this.connect());
        return;
      }

      if (res.statusCode !== 200) {
        this.emit('error', { type: 'error', status: res.statusCode, message: res.statusMessage });
        this.close();
        return;
      }

      this._readyState = EventSourceReadyState.OPEN;

      res.on('close', () => {
        res.removeAllListeners('close');
        res.removeAllListeners('end');
        this.onConnectionClosed();
      });

      res.on('end', () => {
        res.removeAllListeners('close');
        res.removeAllListeners('end');
        this.onConnectionClosed();
      });

      this.emit('open', { type: 'open' });

      // text/event-stream parser adapted from webkit's
      // Source/WebCore/page/EventSource.cpp
      this.isFirstChunk = true;
      this.chunkBuffer = undefined;
      this.startingPosition = 0;
      this.startingFieldLength = -1;
      res.on('data', this.parseDataChunk);
    });

    this.request.on('error', (err) => {
      this.onConnectionClosed(err.message, err);
    });

    if (typeof this.request.setNoDelay === 'function') {
      this.request.setNoDelay(true);
    }
    this.request.end();
  }

  // eslint-disable-next-line class-methods-use-this
  private hasBom(buffer: Buffer): boolean {
    return [239, 187, 191].every((charCode, index) => buffer[index] === charCode);
  }

  private parseDataChunk = (chunk: Buffer): void => {
    this.chunkBuffer = this.chunkBuffer ? Buffer.concat([this.chunkBuffer, chunk]) : chunk;
    if (this.isFirstChunk && this.hasBom(this.chunkBuffer)) {
      this.chunkBuffer = this.chunkBuffer.slice(3);
    }

    this.isFirstChunk = false;
    let position = 0;
    const { length } = this.chunkBuffer;

    while (position < length) {
      if (this.discardTrailingNewline) {
        if (this.chunkBuffer[position] === 10) {
          position += 1;
        }
        this.discardTrailingNewline = false;
      }

      let lineLength = -1;
      let fieldLength = this.startingFieldLength;
      let character: number;

      for (let i = this.startingPosition; lineLength < 0 && i < length; i += 1) {
        character = this.chunkBuffer[i];
        if (character === 58 && fieldLength < 0) {
          fieldLength = i - position;
        } else if (character === 13) {
          this.discardTrailingNewline = true;
          lineLength = i - position;
        } else if (character === 10) {
          lineLength = i - position;
        }
      }

      if (lineLength < 0) {
        this.startingPosition = length - position;
        this.startingFieldLength = fieldLength;
        break;
      } else {
        this.startingPosition = 0;
        this.startingFieldLength = -1;
      }

      this.parseEventStreamLine(this.chunkBuffer, position, fieldLength, lineLength);

      position += lineLength + 1;
    }

    if (position === length) {
      this.chunkBuffer = undefined;
    } else if (position > 0) {
      this.chunkBuffer = this.chunkBuffer.slice(position);
    }
  };

  private parseEventStreamLine(chunkBuffer: Buffer, position: number, fieldLength: number, lineLength: number): void {
    if (lineLength === 0) {
      if (this.data) {
        const type = this.eventName || 'message';
        this.emit(type, {
          data: this.data.slice(0, -1), // remove trailing newline
          lastEventId: this.lastEventId,
          origin: original(this._url),
          type,
        });
        this.data = '';
      }
      this.eventName = '';
    } else if (fieldLength > 0) {
      const noValue = fieldLength < 0;
      let step = 0;
      const field = chunkBuffer.slice(position, position + (noValue ? lineLength : fieldLength)).toString();

      if (noValue) {
        step = lineLength;
      } else if (chunkBuffer[position + fieldLength + 1] !== 32) {
        step = fieldLength + 1;
      } else {
        step = fieldLength + 2;
      }

      const valueLength = lineLength - step;
      const value = chunkBuffer.slice(position + step, position + step + valueLength).toString();

      if (field === 'data') {
        this.data += `${value}\n`;
      } else if (field === 'event') {
        this.eventName = value;
      } else if (field === 'id') {
        this.lastEventId = value;
      } else if (field === 'retry') {
        const retry = parseInt(value, 10);
        if (!Number.isNaN(retry) && this.reconnectionInterval !== 0) {
          this.reconnectionInterval = retry;
        }
      }
    }
  }

  private onConnectionClosed(message?: string, error?: Error): void {
    if (this._readyState === EventSourceReadyState.CLOSED) return;

    this._readyState = EventSourceReadyState.CONNECTING;

    this.emit('error', { type: 'error', error, message });

    if (this.reconnectionInterval === 0) {
      this._readyState = EventSourceReadyState.CLOSED;
      return;
    }

    this._url = this.originalUrl;

    setTimeout(() => {
      if (this._readyState !== EventSourceReadyState.CONNECTING || this.connectionInProgress) return;

      this.connectionInProgress = true;
      this.connect();
    }, this.reconnectionInterval);
  }

  /** Returns the URL providing the event stream. */
  get url(): string {
    return this._url;
  }

  /** Returns true if the credentials mode for connection requests to the URL providing the event stream is set to "include", and false otherwise. */
  get withCredentials(): boolean {
    return this._withCredentials;
  }

  /** Returns the state of this EventSource object's connection. It can have the values described below. */
  get readyState(): EventSourceReadyState {
    return this._readyState;
  }

  set onerror(callback: EventListener) {
    this.removeAllListeners('error');
    if (callback != null) {
      this.on('error', callback);
    }
  }

  set onmessage(callback: EventListener) {
    this.removeAllListeners('message');
    if (callback != null) {
      this.on('message', callback);
    }
  }

  set onopen(callback: EventListener) {
    this.removeAllListeners('open');
    if (callback != null) {
      this.on('open', callback);
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dispatchEvent(event: any): void {
    assert(event.type);
    this.emit(event.type, event.detail || event);
  }

  addEventListener(event: string | symbol, listener: (...args: unknown[]) => void): void {
    this.addListener(event, listener);
  }

  removeEventListener(event: string | symbol, listener: (...args: unknown[]) => void): void {
    this.removeListener(event, listener);
  }

  /** Aborts any instances of the fetch algorithm started for this EventSource object, and sets the readyState attribute to CLOSED. */
  close(): void {
    if (this._readyState === EventSourceReadyState.CLOSED) return;
    this._readyState = EventSourceReadyState.CLOSED;
    if (typeof this.request?.abort === 'function') this.request.abort();
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    if (typeof this.request?.xhr === 'object' && typeof this.request?.xhr.abort === 'function') this.request.xhr.abort();
  }
}

module.exports = EventSource;


lib/eventsource.d.ts
/// <reference types="node" />
import { EventEmitter } from 'events';
import https from 'https';
export declare enum EventSourceReadyState {
    CONNECTING = 0,
    OPEN = 1,
    CLOSED = 2
}
export interface EventSourceInit {
    withCredentials?: boolean;
    reconnectionInterval?: number;
    headers?: {
        [index: string]: string;
    };
    rejectUnauthorized?: boolean;
    proxy?: string;
    https?: https.RequestOptions;
}
export default class EventSource extends EventEmitter {
    readonly CONNECTING = EventSourceReadyState.CONNECTING;
    readonly OPEN = EventSourceReadyState.OPEN;
    readonly CLOSED = EventSourceReadyState.CLOSED;
    static readonly CONNECTING = EventSourceReadyState.CONNECTING;
    static readonly OPEN = EventSourceReadyState.OPEN;
    static readonly CLOSED = EventSourceReadyState.CLOSED;
    private _readyState;
    private originalUrl;
    private _url;
    private _withCredentials;
    private reconnectionInterval;
    private headers;
    private rejectUnauthorized;
    private proxy?;
    private httpsOptions?;
    private request?;
    private connectionInProgress;
    private lastEventId;
    private isFirstChunk;
    private chunkBuffer?;
    private discardTrailingNewline;
    private startingPosition;
    private startingFieldLength;
    private eventName;
    private data;
    constructor(url: string, eventSourceInitDict?: EventSourceInit);
    private connect;
    private hasBom;
    private parseDataChunk;
    private parseEventStreamLine;
    private onConnectionClosed;
    get url(): string;
    get withCredentials(): boolean;
    get readyState(): EventSourceReadyState;
    set onerror(callback: EventListener);
    set onmessage(callback: EventListener);
    set onopen(callback: EventListener);
    dispatchEvent(event: any): void;
    addEventListener(event: string | symbol, listener: (...args: unknown[]) => void): void;
    removeEventListener(event: string | symbol, listener: (...args: unknown[]) => void): void;
    close(): void;
}


lib/eventsource.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSourceReadyState = void 0;
var original_1 = __importDefault(require("original"));
var url_1 = require("url");
var events_1 = require("events");
var https_1 = __importDefault(require("https"));
var http_1 = __importDefault(require("http"));
var assert_1 = __importDefault(require("assert"));
var EventSourceReadyState;
(function (EventSourceReadyState) {
    EventSourceReadyState[EventSourceReadyState["CONNECTING"] = 0] = "CONNECTING";
    EventSourceReadyState[EventSourceReadyState["OPEN"] = 1] = "OPEN";
    EventSourceReadyState[EventSourceReadyState["CLOSED"] = 2] = "CLOSED";
})(EventSourceReadyState = exports.EventSourceReadyState || (exports.EventSourceReadyState = {}));
var EventSource = (function (_super) {
    __extends(EventSource, _super);
    function EventSource(url, eventSourceInitDict) {
        if (eventSourceInitDict === void 0) { eventSourceInitDict = {}; }
        var _a, _b, _c;
        var _this = _super.call(this) || this;
        _this.CONNECTING = EventSourceReadyState.CONNECTING;
        _this.OPEN = EventSourceReadyState.OPEN;
        _this.CLOSED = EventSourceReadyState.CLOSED;
        _this._readyState = EventSourceReadyState.CONNECTING;
        _this.connectionInProgress = false;
        _this.lastEventId = '';
        _this.isFirstChunk = false;
        _this.discardTrailingNewline = false;
        _this.startingPosition = 0;
        _this.startingFieldLength = -1;
        _this.eventName = '';
        _this.data = '';
        _this.parseDataChunk = function (chunk) {
            _this.chunkBuffer = _this.chunkBuffer ? Buffer.concat([_this.chunkBuffer, chunk]) : chunk;
            if (_this.isFirstChunk && _this.hasBom(_this.chunkBuffer)) {
                _this.chunkBuffer = _this.chunkBuffer.slice(3);
            }
            _this.isFirstChunk = false;
            var position = 0;
            var length = _this.chunkBuffer.length;
            while (position < length) {
                if (_this.discardTrailingNewline) {
                    if (_this.chunkBuffer[position] === 10) {
                        position += 1;
                    }
                    _this.discardTrailingNewline = false;
                }
                var lineLength = -1;
                var fieldLength = _this.startingFieldLength;
                var character = void 0;
                for (var i = _this.startingPosition; lineLength < 0 && i < length; i += 1) {
                    character = _this.chunkBuffer[i];
                    if (character === 58 && fieldLength < 0) {
                        fieldLength = i - position;
                    }
                    else if (character === 13) {
                        _this.discardTrailingNewline = true;
                        lineLength = i - position;
                    }
                    else if (character === 10) {
                        lineLength = i - position;
                    }
                }
                if (lineLength < 0) {
                    _this.startingPosition = length - position;
                    _this.startingFieldLength = fieldLength;
                    break;
                }
                else {
                    _this.startingPosition = 0;
                    _this.startingFieldLength = -1;
                }
                _this.parseEventStreamLine(_this.chunkBuffer, position, fieldLength, lineLength);
                position += lineLength + 1;
            }
            if (position === length) {
                _this.chunkBuffer = undefined;
            }
            else if (position > 0) {
                _this.chunkBuffer = _this.chunkBuffer.slice(position);
            }
        };
        _this.originalUrl = url;
        _this._url = url;
        _this._withCredentials = (_a = eventSourceInitDict.withCredentials) !== null && _a !== void 0 ? _a : false;
        _this.reconnectionInterval = (_b = eventSourceInitDict.reconnectionInterval) !== null && _b !== void 0 ? _b : 1000;
        _this.headers = __assign({ 'Cache-Control': 'no-cache', Accept: 'text/event-stream' }, eventSourceInitDict.headers);
        _this.rejectUnauthorized = (_c = eventSourceInitDict.rejectUnauthorized) !== null && _c !== void 0 ? _c : true;
        _this.proxy = eventSourceInitDict.proxy;
        _this.httpsOptions = eventSourceInitDict.https;
        if (_this.headers['Last-Event-ID']) {
            _this.lastEventId = _this.headers['Last-Event-ID'];
            delete _this.headers['Last-Event-ID'];
        }
        process.nextTick(function () { return _this.connect(); });
        return _this;
    }
    EventSource.prototype.connect = function () {
        var _this = this;
        var _a;
        var reqOptions = url_1.parse(this._url);
        var isSecure = reqOptions.protocol === 'https:';
        reqOptions.headers = this.headers;
        if (this.lastEventId) {
            reqOptions.headers['Last-Event-ID'] = this.lastEventId;
        }
        reqOptions.rejectUnauthorized = this.rejectUnauthorized;
        if (this.proxy) {
            var proxy = url_1.parse(this.proxy);
            isSecure = proxy.protocol === 'https:';
            reqOptions.protocol = isSecure ? 'https:' : 'http:';
            reqOptions.path = this._url;
            reqOptions.headers.Host = (_a = reqOptions.host) !== null && _a !== void 0 ? _a : undefined;
            reqOptions.hostname = proxy.hostname;
            reqOptions.host = proxy.host;
            reqOptions.port = proxy.port;
        }
        if (this.httpsOptions) {
            reqOptions = __assign(__assign({}, reqOptions), this.httpsOptions);
        }
        if (this._withCredentials) {
            reqOptions.withCredentials = this._withCredentials;
        }
        this.request = (isSecure ? https_1.default : http_1.default).request(reqOptions, function (res) {
            _this.connectionInProgress = false;
            if (res.statusCode && res.statusCode >= 500) {
                _this.emit('error', { type: 'error', status: res.statusCode, message: res.statusMessage });
                _this.onConnectionClosed();
                return;
            }
            if (res.statusCode === 301 || res.statusCode === 302 || res.statusCode === 307) {
                if (!res.headers.location) {
                    _this.emit('error', { type: 'error', status: res.statusCode, message: 'Server sent redirect response without Location header.' });
                    return;
                }
                if (res.statusCode === 301)
                    _this.originalUrl = res.headers.location;
                _this._url = res.headers.location;
                process.nextTick(function () { return _this.connect(); });
                return;
            }
            if (res.statusCode !== 200) {
                _this.emit('error', { type: 'error', status: res.statusCode, message: res.statusMessage });
                _this.close();
                return;
            }
            _this._readyState = EventSourceReadyState.OPEN;
            res.on('close', function () {
                res.removeAllListeners('close');
                res.removeAllListeners('end');
                _this.onConnectionClosed();
            });
            res.on('end', function () {
                res.removeAllListeners('close');
                res.removeAllListeners('end');
                _this.onConnectionClosed();
            });
            _this.emit('open', { type: 'open' });
            _this.isFirstChunk = true;
            _this.chunkBuffer = undefined;
            _this.startingPosition = 0;
            _this.startingFieldLength = -1;
            res.on('data', _this.parseDataChunk);
        });
        this.request.on('error', function (err) {
            _this.onConnectionClosed(err.message, err);
        });
        if (typeof this.request.setNoDelay === 'function') {
            this.request.setNoDelay(true);
        }
        this.request.end();
    };
    EventSource.prototype.hasBom = function (buffer) {
        return [239, 187, 191].every(function (charCode, index) { return buffer[index] === charCode; });
    };
    EventSource.prototype.parseEventStreamLine = function (chunkBuffer, position, fieldLength, lineLength) {
        if (lineLength === 0) {
            if (this.data) {
                var type = this.eventName || 'message';
                this.emit(type, {
                    data: this.data.slice(0, -1),
                    lastEventId: this.lastEventId,
                    origin: original_1.default(this._url),
                    type: type,
                });
                this.data = '';
            }
            this.eventName = '';
        }
        else if (fieldLength > 0) {
            var noValue = fieldLength < 0;
            var step = 0;
            var field = chunkBuffer.slice(position, position + (noValue ? lineLength : fieldLength)).toString();
            if (noValue) {
                step = lineLength;
            }
            else if (chunkBuffer[position + fieldLength + 1] !== 32) {
                step = fieldLength + 1;
            }
            else {
                step = fieldLength + 2;
            }
            var valueLength = lineLength - step;
            var value = chunkBuffer.slice(position + step, position + step + valueLength).toString();
            if (field === 'data') {
                this.data += value + "\n";
            }
            else if (field === 'event') {
                this.eventName = value;
            }
            else if (field === 'id') {
                this.lastEventId = value;
            }
            else if (field === 'retry') {
                var retry = parseInt(value, 10);
                if (!Number.isNaN(retry) && this.reconnectionInterval !== 0) {
                    this.reconnectionInterval = retry;
                }
            }
        }
    };
    EventSource.prototype.onConnectionClosed = function (message, error) {
        var _this = this;
        if (this._readyState === EventSourceReadyState.CLOSED)
            return;
        this._readyState = EventSourceReadyState.CONNECTING;
        this.emit('error', { type: 'error', error: error, message: message });
        if (this.reconnectionInterval === 0) {
            this._readyState = EventSourceReadyState.CLOSED;
            return;
        }
        this._url = this.originalUrl;
        setTimeout(function () {
            if (_this._readyState !== EventSourceReadyState.CONNECTING || _this.connectionInProgress)
                return;
            _this.connectionInProgress = true;
            _this.connect();
        }, this.reconnectionInterval);
    };
    Object.defineProperty(EventSource.prototype, "url", {
        get: function () {
            return this._url;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EventSource.prototype, "withCredentials", {
        get: function () {
            return this._withCredentials;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EventSource.prototype, "readyState", {
        get: function () {
            return this._readyState;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EventSource.prototype, "onerror", {
        set: function (callback) {
            this.removeAllListeners('error');
            if (callback != null) {
                this.on('error', callback);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EventSource.prototype, "onmessage", {
        set: function (callback) {
            this.removeAllListeners('message');
            if (callback != null) {
                this.on('message', callback);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EventSource.prototype, "onopen", {
        set: function (callback) {
            this.removeAllListeners('open');
            if (callback != null) {
                this.on('open', callback);
            }
        },
        enumerable: false,
        configurable: true
    });
    EventSource.prototype.dispatchEvent = function (event) {
        assert_1.default(event.type);
        this.emit(event.type, event.detail || event);
    };
    EventSource.prototype.addEventListener = function (event, listener) {
        this.addListener(event, listener);
    };
    EventSource.prototype.removeEventListener = function (event, listener) {
        this.removeListener(event, listener);
    };
    EventSource.prototype.close = function () {
        var _a, _b, _c;
        if (this._readyState === EventSourceReadyState.CLOSED)
            return;
        this._readyState = EventSourceReadyState.CLOSED;
        if (typeof ((_a = this.request) === null || _a === void 0 ? void 0 : _a.abort) === 'function')
            this.request.abort();
        if (typeof ((_b = this.request) === null || _b === void 0 ? void 0 : _b.xhr) === 'object' && typeof ((_c = this.request) === null || _c === void 0 ? void 0 : _c.xhr.abort) === 'function')
            this.request.xhr.abort();
    };
    EventSource.CONNECTING = EventSourceReadyState.CONNECTING;
    EventSource.OPEN = EventSourceReadyState.OPEN;
    EventSource.CLOSED = EventSourceReadyState.CLOSED;
    return EventSource;
}(events_1.EventEmitter));
exports.default = EventSource;
module.exports = EventSource;
